### PR TITLE
Fault

### PR DESCRIPTION
**Pull request by bitflipp**

### CODE REVIEW

File: runner_test.go [hunk 1]
```diff
@@ -99,3 +169,176 @@ func TestRunnerPersistence(t *testing.T) {
...
+		testNoError(t, rn.initialize())
+		go func() {
+			testNoError(t, r.worker(false))
+		}()
+		time.Sleep(100 * time.Millisecond)
+		testSendInterrupt()
+		time.Sleep(100 * time.Millisecond)
+	}
+
+	ts([]string{"file", "test/file"})
```
Comment (Top-Ranger) [inline runner_test.go:332]: Maybe use better file name e.g. "emptyFile" or "empty"

File: main.go [hunk 2]
```diff
@@ -20,15 +20,6 @@ func main() {
 	cfp := flag.String("c", "./gerberos.toml", "Path to TOML configuration file")
 	flag.Parse()
 
-	// Already running instance?
```
Comment (Top-Ranger) [inline main.go:23]: Is there a replacement or do we just drop this feature?
I don't mind if we drop it

**Review by Top-Ranger — CHANGES_REQUESTED** [review]
There is an error in the tests:
```
=== RUN   TestRunnerPerformActionFlaky
2022/03/20 18:12:35 warning: not persisting ipsets
    testing.go:58: unexpected error: failed to initialize backend: failed to create ipsets: failed to create ipset "gerberos4": ipset v7.10: Set cannot be created: set with the same name already exists
```
Most likely the backend was not finalised in a prior run. Please fix.

**Comment by Top-Ranger** [comment]
An additional note: the nft sets "gerberos4" and "gerberos6" are existing after the test failed

File: main.go (hunk 2 above)
Comment (bitflipp) [reply main.go:23]: Let's drop it. systemd (or similar) will prevent accidental multiple instances.

File: runner_test.go (hunk 1 above)
Comment (bitflipp) [reply runner_test.go]: Revised.

File: go.mod [hunk 3]
```diff
@@ -1,5 +1,5 @@
 module github.com/bitflipp/gerberos
 
-go 1.17
+go 1.18
```
[the remark this answers is not in the document]
Comment (bitflipp) [reply go.mod:3]: Compiler and linker improvements. You can still use the 1.17 toolchain.

File: backends.go [hunk 4]
```diff
@@ -315,19 +310,18 @@ func (b *nftBackend) Ban(ip string, ipv6 bool, d time.Duration) error {
...
+				// v0.9.3, this is needed.
 				return nil
 			}
 			return fmt.Errorf(`failed to add element to set "%s": %s`, b.set6Name, s)
 		}
 	} else {
-		if s, ec, err := execute(nil, "nft", "add", "element", "ip", b.table4Name, b.set4Name, fmt.Sprintf("{ %s timeout %ds }", ip, ds)); err != nil {
+		if s, ec, err := b.runner.executor.execute("nft", "add", "element", "ip", b.table4Name, b.set4Name, fmt.Sprintf("{ %s timeout %ds }", ip, ds)); err != nil {
 			if ec == 1 {
-				// This ip is probably already in set. Ignore the error.
```
[the remark this answers is not in the document]
Comment (bitflipp) [reply backends.go:329]: The real question here is...Why was no deduplication effort made?

**Review by Top-Ranger — APPROVED** [review]

**Review by Top-Ranger — APPROVED** [review]

**Pull request merged**